coala_langserver/langserver.py

```diff
@@ -102,7 +93,46 @@ def send_diagnostics(self, path, diagnostics):
             'uri': 'file://{0}'.format(path),
             'diagnostics': _diagnostics,
         }
-        self.send_notification('textDocument/publishDiagnostics', params)
+        self._endpoint.notify('textDocument/publishDiagnostics', params=params)
+
+
+def start_tcp_lang_server(bind_addr, port, handler_class):
+    if not issubclass(handler_class, LangServer):
+        raise ValueError('Handler class must be an instance of LangServer')
+
+    # Construct a custom wrapper class around the user's handler_class
+    wrapper_class = type(
+        handler_class.__name__ + 'Handler',
+        (_StreamHandlerWrapper,),
+        {'DELEGATE_CLASS': handler_class}
+    )
+
+    try:
+        server = socketserver.TCPServer((bind_addr, port), wrapper_class)
+    except Exception as e:
+        log('Fatal Exception: {}'.format(e))
+        sys.exit(1)
+    try:
+        log('Serving {} on ({}, {})'.format(
+            handler_class.__name__, bind_addr, port))
+        server.serve_forever()
+    finally:
+        log('Shutting down')
+        server.server_close()
+
+
+def start_io_lang_server(handler_class, rfile=None, wfile=None):
+    if rfile is None:
+        rfile = sys.stdin.buffer
+    if wfile is None:
+        wfile = sys.stdout.buffer
+
+    if not issubclass(handler_class, LangServer):
```

Comment:
declare arg `handler_class: LangServer`

and coala has a decorator to enforce type annotations.

you are in a Py3 only project now ;-)